drivers/dma/dma_sedi.c

```diff
@@ -0,0 +1,582 @@
+	for (int i = 0; i < DMA_CHANNEL_NUM; i++) {
+		k_sem_init(&data->sema[i], 1, 1);
+	}
+	return 0;
+}
+
+static int dma_sedi_init(const struct device *dev)
+{
+	const struct dma_sedi_config_info *const info = DEV_CFG(dev);
+	sedi_dma_t dma_dev = info->instance;
+	uint32_t irq_flag = 0;
+
+	if (!sedi_dev_is_self_owned(SEDI_DEVID_DMA0 + dma_dev)) {
+		return -ENODEV;
+	}
+
+	_impl_dma_sedi_init(dev);
+	switch (dma_dev) {
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay)
+	case SEDI_DMA_0:
+#if DT_INST_IRQ_HAS_CELL(0, sense)
+		irq_flag = DT_INST_IRQ(0, sense);
+#endif
+		IRQ_CONNECT(DT_INST_IRQN(0),
+			    DT_INST_IRQ(0, priority),
+			    dma_isr, (void *)SEDI_DMA_0, irq_flag);
+		irq_enable(DT_INST_IRQN(0));
+		break;
+#endif
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(1), okay)
+	case SEDI_DMA_1:
+#if DT_INST_IRQ_HAS_CELL(1, sense)
+		irq_flag = DT_INST_IRQ(1, sense);
+#endif
+		IRQ_CONNECT(DT_INST_IRQN(1),
+			    DT_INST_IRQ(1, priority),
+			    dma_isr, (void *)SEDI_DMA_1, irq_flag);
+		irq_enable(DT_INST_IRQN(1));
+		break;
+#endif
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(2), okay)
+	case SEDI_DMA_2:
+#if DT_INST_IRQ_HAS_CELL(2, sense)
+		irq_flag = DT_INST_IRQ(2, sense);
+#endif
+		IRQ_CONNECT(DT_INST_IRQN(2),
+			    DT_INST_IRQ(2, priority),
+			    dma_isr, (void *)SEDI_DMA_2, irq_flag);
+		irq_enable(DT_INST_IRQN(2));
+		break;
+#endif
+	default:
+		return -EIO;
+	}
+#if defined(CONFIG_PM_DEVICE)
+	struct dma_sedi_driver_data *const data = DEV_DATA(dev);
+
+	data->power_status = PM_DEVICE_STATE_ACTIVE;
+#endif
+	return 0;
+}
+
+#ifdef CONFIG_PM_DEVICE
+
+static bool is_dma_busy(sedi_dma_t dev)
+{
+	sedi_dma_status_t chn_status;
+
+	for (int chn = 0; chn < DMA_CHANNEL_NUM; chn++) {
+		sedi_dma_get_status(dev, chn, &chn_status);
+		if (chn_status.busy == 1) {
+			return true;
+		}
+	}
+	return false;
+}
+
+static int dma_change_device_power(const struct device *dev,
+		enum pm_device_action action)
+{
+	struct dma_sedi_driver_data *const data = DEV_DATA(dev);
+	const struct dma_sedi_config_info *const info = DEV_CFG(dev);
+	sedi_dma_t dma_dev = info->instance;
+	int ret;
+
+	sedi_power_state_t state;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		state = SEDI_POWER_FULL;
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		if (is_dma_busy(dma_dev)) {
+			return -EBUSY;
+		}
+		state = SEDI_POWER_SUSPEND;
+		break;
+
+	default:
+		return -ENOTSUP;
+	}
+
+	for (uint8_t chn = 0; chn < DMA_CHANNEL_NUM; chn++) {
+		ret = sedi_dma_set_power(dma_dev, chn, state);
+		if (ret != SEDI_DRIVER_OK) {
+			return -EIO;
+		}
+	}
+
+	data->power_status = action;
+	return 0;
+}
+
+static int dma_sedi_device_ctrl(const struct device *dev,
+		enum pm_device_action action)
+{
+	int ret = 0;
+
+	ret = dma_change_device_power(dev, action);
+
+	return ret;
+}
+
+#endif
+
+#define DMA_DEVICE_INIT_SEDI(n) \
+	static struct dma_sedi_driver_data dma_sedi_dev_data_##n; \
+	static const struct dma_sedi_config_info dma_sedi_config_data_##n = { \
+		.instance = SEDI_DMA_##n, \
```

Comment:
add "peripheral-id" in dts yaml, and use it instead of "instance"

Reply:
Done